runtime/script_loader.cpp

```diff
@@ -119,8 +126,28 @@ JSObject* module_resolve_hook(JSContext* cx, HandleValue referencingPrivate,
     return nullptr;
   }
   JS::CompileOptions opts(cx, *COMPILE_OPTS);
-  opts.setFileAndLine(path.get(), 1);
-  auto resolved_path = resolve_path(path.get(), BASE_PATH);
+  opts.setFileAndLine(strip_base(path.get(), BASE_PATH), 1);
+
+  RootedObject info(cx, &referencingPrivate.toObject());
+  RootedValue parent_path_val(cx);
+  if (!JS_GetProperty(cx, info, "path", &parent_path_val)) {
+    return nullptr;
+  }
+  if (!parent_path_val.isString()) {
+    return nullptr;
+  }
+  RootedString parent_path(cx, parent_path_val.toString());
+  size_t parent_len = JS_GetStringLength(parent_path) * 3 + 1;
+  char* buf(js_pod_malloc<char>(parent_len));
+  auto result = JS_EncodeStringToUTF8BufferPartial(cx, parent_path, mozilla::Span<char>(buf, buf + parent_len));
+  if (result.isNothing()) {
+    js_free(buf);
+    return nullptr;
+  }
+  size_t written = std::get<1>(result.extract());
+  buf[written] = '\0';
+  char* resolved_path = resolve_path(path.get(), buf);
+  js_free(buf);
```

Comment:
It'd be nice if we could simplify this a bit, and I see a few different options for doing so (vaguely in descending order of neatness, I think, though I find myself waffling between the first two):

1. We could (in addition to storing it as a JS string under `path` as right now) store the path as a private value that holds the `char*` so it doesn't need to be re-encoded for all imports.
2. We could retain the base path in a simple linked list that gets added to every time we start parsing a file, and popped from when we're done with it, at which point we can free the chars.
3. We could clean up the code by using `core::encode` instead of implementing encoding here.

Separately from all that, we could pass the base path's length into `resolve_path` so it doesn't need to be re-derived there.

Reply:
I've added (3), and also added the explicit length pass.